src/nvim/buffer.c

```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
+    // Note: This is not needed if `str` is pointing at `tmp`
+    if (opt == STL_VIM_EXPR) {
       xfree(str);
+    }
 
     if (num >= 0 || (!itemisflag && str && *str))
-      prevchar_isflag = FALSE;              /* Item not NULL, but not a flag */
+      prevchar_isflag = false;              /* Item not NULL, but not a flag */
+
+    // Item processed, move to the next
     curitem++;
   }
-  *p = NUL;
-  itemcnt = curitem;
 
-  if (usefmt != fmt)
+  // Null terminate the output buffer
+  *out_p = NUL;
+  int itemcnt = curitem;
+
+  // Free the format buffer if we allocated it internally
+  if (usefmt != fmt) {
     xfree(usefmt);
+  }
 
-  width = vim_strsize(out);
+  int width = vim_strsize(out);
   if (maxwidth > 0 && width > maxwidth) {
-    /* Result is too long, must truncate somewhere. */
-    l = 0;
-    if (itemcnt == 0)
-      s = out;
-    else {
-      for (; l < itemcnt; l++)
-        if (item[l].type == Trunc) {
-          /* Truncate at %< item. */
-          s = item[l].start;
+    // Result is too long, must truncate somewhere.
+    int item_idx = 0;
+    char_u *trunc_p;
+
+    // If there are no items, truncate from beginning
+    if (itemcnt == 0) {
+      trunc_p = out;
+
+    // Otherwise, look for the truncation item
+    } else {
+      // Default to truncating at the first item
+      trunc_p = item[0].start;
+      item_idx = 0;
+
+      for (int i = 0; i < itemcnt; i++)
+        if (item[i].type == Trunc) {
+          // Truncate at %< item.
+          trunc_p = item[i].start;
+          item_idx = i;
           break;
         }
-      if (l == itemcnt) {
-        /* No %< item, truncate first item. */
-        s = item[0].start;
-        l = 0;
-      }
     }
 
-    if (width - vim_strsize(s) >= maxwidth) {
-      /* Truncation mark is beyond max length */
+    // If the truncation point we found is beyond the maximum
+    // length of the string, truncate the end of the string.
+    if (width - vim_strsize(trunc_p) >= maxwidth) {
+      // If we are using a multi-byte encoding, walk from the beginning of the
+      // string to find the last character that will fit.
       if (has_mbyte) {
-        s = out;
+        trunc_p = out;
         width = 0;
         for (;; ) {
-          width += ptr2cells(s);
+          width += ptr2cells(trunc_p);
           if (width >= maxwidth)
             break;
-          s += (*mb_ptr2len)(s);
+
+          // Note: Only advance the pointer if the next
+          //       character will fit in the available output space
+          trunc_p += (*mb_ptr2len)(trunc_p);
         }
-        /* Fill up for half a double-wide character. */
-        while (++width < maxwidth)
-          *s++ = fillchar;
-      } else
-        s = out + maxwidth - 1;
-      for (l = 0; l < itemcnt; l++)
-        if (item[l].start > s)
+        // Fill up for half a double-wide character.
+        // XXX : This seems impossible given the exit condition above?
```

Comment:
Could have been left-over from Vim source before removing `#ifdef` blocks which maybe confused the original author(s). Would be worth peeking at the Vim source here.